L42-python-metadata-flags.md

```diff
@@ -0,0 +1,150 @@
+
+Strictly speaking, the `_ChannelReadyFuture` works in most cases. Though it is fungible to 'wait-for-ready', this interface has to be maintained in the near future.
+
+### Desired 'wait-for-ready' behavior
+
+The desired behavior will be able to config each RPC call with different settings. Currently, Python is using keyword argument for call level configuration, so this feature will use keyword argument as well. The keyword argument mechanism also enables developers to configure certain calls with one setting in many ways (like expand a dictionary-like object every time).
+
+In other languages, there is a few difference in implementation. They prefer to have a separate class for call configuration ('ClientContext' in C++, 'CallOptions' in Java & Go), then pass it to every RPC invocation. The new design can be equivalence in behavior.
+
+### The 3-states switch of 'wait-for-ready' in C-Core
+
+The implementation of 'wait-for-ready' flags is not as simple as a one bit switch. It contains two bits one is `GRPC_INITIAL_METADATA_WAIT_FOR_READY` another one is `GRPC_INITIAL_METADATA_WAIT_FOR_READY_EXPLICITLY_SET`. Although two bits can represent four states, three of them are valid.
+
+Assume `flag` is the flag variable we are manipulating.
+
+If the developer wants to use default value, then `flag = flag & ~WAIT_FOR_READY & ~WAIT_FOR_READY_EXPLICITLY_SET` (both 0).
+
+If the developer doesn't want to 'wait-for-ready', then `flag = flag & ~WAIT_FOR_READY & WAIT_FOR_READY_EXPLICITLY_SET` (0, 1).
+
+If the developer wants to 'wait-for-ready', then `flag = flag & WAIT_FOR_READY & WAIT_FOR_READY_EXPLICITLY_SET` (1, 1).
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+* Add an optional `wait_for_ready` variable to `MultiCallable` classes initialization methods. Default `None`, accept `bool`.
+* Import initial metadata flags constants from `grpc_types.h` to `grpc.pxi`.
+
+### DEMO snippets 
+
+```Python
+# Per RPC level
+stub = ...Stub(...)
+
+stub.initial_rpc_call(..., wait_for_ready=True)
+stub.important_transaction_1(..., wait_for_ready=True)
+stub.unimportant_transaction_2(...)
+stub.important_transaction_3(..., wait_for_ready=True)
+stub.unimportant_transaction_4(...)
+# The unimportant transactions can be status report, or health check, etc.
+```
+
+```Python
+# Per RPC level with configuration for subsets
+stub = ...Stub(...)
+
+important_call_options = {'wait_for_ready': True, 'timeout': 4000}
+failfast_call_options = {'wait_for_ready': False}
+
+stub.initial_rpc_call(..., **important_call_options)
+stub.important_transaction_1(..., **important_call_options)
+stub.unimportant_transaction_2(..., **failfast_call_options)
+stub.important_transaction_3(..., **important_call_options)
+stub.unimportant_transaction_4(..., **failfast_call_options)
+```
+
+```Python
+# Channel level - Invalid Design
+channel = grpc.insecure_channel(..., wait_for_ready=True)
```

Comment:
nit: because this channel setting is under demo snippets (even though it's under the `# Channel level - Invalid Design` comment), it seems not totally obvious that this API was something considered but not supported (correct?). Suggest making the comment more explicit and/or moving this to a different section

Reply:
Will do, good point!

Reply:
I would go further and just delete this altogether. All this seems like it will do is encourage people to copy and paste an example that won't work.